### PR TITLE
fix: handle `build.format` links in sidebar

### LINKED ISSUE #180: Setting Astro `build.format` config breaks sidebar links
### What version of `starlight` are you using?

0.1.2

### What version of `astro` are you using?

2.6.1

### What package manager are you using?

npm

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

**Expected**: sidebar links should not have a trailing slash when setting [Astro's build.format config](https://docs.astro.build/en/reference/configuration-reference/#buildformat) to `file`

**Actual**: they have a trailing slash, which causes the links to 404 in production builds

**How to reproduce**:

1. Clone the `build-format-file` branch of [this repo](https://github.com/mergentlabs/docs)
2. Comment out [line 8 of the Astro config](https://github.com/mergentlabs/docs/blob/main/astro.config.mjs#L8) that sets the base path to `/docs` (sorry, it's a temporary workaround for GitHub pages)
3. `npm run dev` to see the trailing `/` OR `npm run build && npm run preview` to get the 404s
4. Click any link in the sidebar
5. Smile at everyone's favorite sad piece of paper :)
Screenshot 2023-06-08 at 5 40 36 PM

### Link to Minimal Reproducible Example

https://github.com/mergentlabs/docs/tree/build-format-file

### Participation

- [X] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by kevinzunigacuellar**
#### What kind of changes does this PR include?

- Changes to Starlight code

#### Description

- Closes #180
- Implemented `formatPath` function to adjust path formats according to project configurations.

- Defined conventions for formatting paths:
  - For `build.format: "file"`, the path will conclude with a ".html" extension. For instance, "/reference/" will transform into "/reference.html".
  - For `build.format: "directory"`, the path will include a trailing slash. For instance, "/reference.html" will change to "/reference/".

### CODE REVIEW

File: packages/starlight/integrations/virtual-user-config.ts [hunk 1]
```diff
@@ -10,15 +10,15 @@ function resolveVirtualModuleId<T extends string>(id: T): `\0${T}` {
...
+	{ root, srcDir, build }: Pick<AstroConfig, 'root' | 'srcDir' | 'build' >
 ): NonNullable<ViteUserConfig['plugins']>[number] {
 	const resolveId = (id: string) =>
 		JSON.stringify(id.startsWith('.') ? resolve(fileURLToPath(root), id) : id);
 
 	/** Map of virtual module names to their code contents as strings. */
 	const modules = {
 		'virtual:starlight/user-config': `export default ${JSON.stringify(opts)}`,
-		'virtual:starlight/project-context': `export default ${JSON.stringify({ root, srcDir })}`,
+		'virtual:starlight/project-context': `export default ${JSON.stringify({ root, srcDir, buildFormat: build.format })}`,
```
Comment (delucis) [inline packages/starlight/integrations/virtual-user-config.ts:21]: I’d be tempted to keep this mirroring Astro’s config shape 1:1:

```suggestion
		'virtual:starlight/project-context': `export default ${JSON.stringify({ root, srcDir, build: { format: build.format } })}`,
```

File: packages/starlight/virtual.d.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@ declare module 'virtual:starlight/user-config' {
 	export default Config;
 }
 declare module 'virtual:starlight/project-context' {
-	export default { root: string, srcDir: string };
+	export default { root: string, srcDir: string, buildFormat: import('astro').AstroConfig['build']['format'] };
```
Comment (delucis) [inline packages/starlight/virtual.d.ts:6]:
```suggestion
	export default { root: string, srcDir: string, build: Pick<AstroConfig['build'], 'format'> };
```

**Review by delucis — COMMENTED** [review]
Thanks @kevinzunigacuellar! I see `build.format: 'file'` in the docs to test, but looks like the sidebar in the deploy preview does not use `.html`. Any idea why?

File: packages/starlight/__tests__/test-config.ts [hunk 3]
```diff
@@ -4,11 +4,26 @@ import { getViteConfig } from 'astro/config';
...
+export function defineVitestConfigBuildFormatFile(config: z.input<typeof StarlightConfigSchema>) {
+	const root = new URL('./', import.meta.url);
+	const srcDir = new URL('./src/', root);
+	const build = {
+		format: "file"
+	} as AstroConfig['build']
+	return getViteConfig({
+		plugins: [vitePluginStarlightUserConfig(StarlightConfigSchema.parse(config), { root, srcDir, build })],
 	});
 }
```
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply packages/starlight/__tests__/test-config.ts:29]: Yes, I wasn't really sure how to test this without copying a directory full of tests. I will try this.

**Comment by kevinzunigacuellar** [comment]
> I see build.format: 'file' in the docs to test, but looks like the sidebar in the deploy preview does not use .html. Any idea why?

Checking the build output it looks that all the href in the sidebar have a .html append to the end. I think it's one of netlify features to strip the .html from the deployed files ([Pretty URL’s](https://docs.netlify.com/configure-builds/file-based-configuration/#pretty-urls))

**Comment by delucis** [comment]
> Checking the build output it looks that all the href in the sidebar have a .html append to the end. I think it's one of netlify features to strip the .html from the deployed files ([Pretty URL’s](https://docs.netlify.com/configure-builds/file-based-configuration/#pretty-urls))

Ahhhhh! Thank you Netlify 🖖

**Comment by kevinzunigacuellar** [comment]
I had to make a few assertions for `AstroConfig['build']` because Partial only makes the top level properties optional and it expects a fully defined build property to be passed to the user config plugin.

File: packages/starlight/integrations/virtual-user-config.ts [hunk 4]
```diff
@@ -10,15 +10,15 @@ function resolveVirtualModuleId<T extends string>(id: T): `\0${T}` {
 /** Vite plugin that exposes Starlight user config and project context via virtual modules. */
 export function vitePluginStarlightUserConfig(
 	opts: StarlightConfig,
-	{ root, srcDir }: Pick<AstroConfig, 'root' | 'srcDir'>
+	{ root, srcDir, build }: Pick<AstroConfig, 'root' | 'srcDir' | 'build' >
```
Comment (delucis) [inline packages/starlight/integrations/virtual-user-config.ts:13]: Ah yeah, good point about the assertions. They’re not the end of the world, but maybe we can type more accurately what we actually need here, so we can avoid them:

```suggestion
	{ root, srcDir, build }: Pick<AstroConfig, 'root' | 'srcDir'> & { build: Pick<AstroConfig['build'], 'format'> }
```

File: packages/starlight/__tests__/test-config.ts [hunk 5]
```diff
@@ -4,11 +4,19 @@ import { getViteConfig } from 'astro/config';
...
 import { StarlightConfigSchema } from '../utils/user-config';
+import type { AstroConfig } from 'astro';
+
+export function defineVitestConfig(
+	config: z.input<typeof StarlightConfigSchema>,
+	opts?: Partial<AstroConfig>
+) {
+	const root = opts?.root ?? new URL('./', import.meta.url);
+	const srcDir = opts?.srcDir ?? new URL('./src/', root);
+	const build = opts?.build ?? ({ format: 'directory' } as AstroConfig['build']);
```
Comment (delucis) [inline packages/starlight/__tests__/test-config.ts:15]: Should allow us to set this without the assertion I think?

```suggestion
	const build = opts?.build ?? { format: 'directory' };
```

File: packages/starlight/__tests__/test-config.ts [hunk 6]
```diff
@@ -4,11 +4,19 @@ import { getViteConfig } from 'astro/config';
 import type { z } from 'astro/zod';
 import { vitePluginStarlightUserConfig } from '../integrations/virtual-user-config';
 import { StarlightConfigSchema } from '../utils/user-config';
+import type { AstroConfig } from 'astro';
+
+export function defineVitestConfig(
+	config: z.input<typeof StarlightConfigSchema>,
+	opts?: Partial<AstroConfig>
```
Comment (delucis) [inline packages/starlight/__tests__/test-config.ts:11]: And then same type here should avoid the assertions in test config:

```suggestion
	opts?: { build: Pick<AstroConfig['build'], 'format'> }
```

File: packages/starlight/__tests__/sidebar-file-build/vitest.config.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+			{
+				label: 'API v1',
+				autogenerate: { directory: '/api/v1/' },
+			},
+		],
+	},
+	{
+		build: {
+			format: 'file',
+		} as AstroConfig['build'],
```
Comment (delucis) [inline packages/starlight/__tests__/sidebar-file-build/vitest.config.ts:50]:
```suggestion
		},
```

File: packages/starlight/utils/navigation.ts [hunk 8]
```diff
@@ -134,8 +135,19 @@ function makeLink(
...
+		return { type: 'link', label, href, isCurrent: false, badge, attrs: attrs ?? {} };
+
+	href = pathWithBase(href);
+	const isCurrent = href === ensureTrailingSlash(stripExtension(currentPathname));
+
+	if (project.build.format === 'file') {
+		// edge case for index.html
+		if (href === '/') href = '/index/';
+		href = stripTrailingSlash(href) + '.html';
+	}
```
Comment (delucis) [inline packages/starlight/utils/navigation.ts:149]: Is there a way to extract some of this into a reusable function that formats any relative path according to Astro’s config?

We’ll need to do the same for the `<SiteTitle />` component which currently does this:

https://github.com/withastro/starlight/blob/5499c2524ca958c6a59581eaf6184e14c4fccecf/packages/starlight/components/SiteTitle.astro#L7

Would also make it easier to test because then we can just have a bunch of tests for that function passing different arguments and checking it returns what we expect.

**Review by delucis — COMMENTED** [review]
Thanks for cleaning up that test config helper. I left a suggestion to avoid the type assertions and then one question.

File: packages/starlight/utils/navigation.ts (hunk 8 above)
Comment (kevinzunigacuellar) [reply packages/starlight/utils/navigation.ts]: I didn't know it was used somewhere else. Yes, I can give it a shot.

**Comment by kevinzunigacuellar** [comment]
I made a function called `formatPathFromConfig` that formats any given path using the astro config options (base and build.format). I am not sure where to put it for now its just on navigation utils. And I wrote a few tests with different edge cases. What do you think?

**Comment by delucis** [comment]
Nice work!

> I am not sure where to put it for now its just on navigation utils.

Might make sense to move it to its own file. We do have `path.ts` and `base.ts`, but those both avoid relying on the virtual imports.

The other option would be to refactor it slightly to take the required config as an argument and put it together in `path.ts`:

```ts
/** Add `.html` extension to a path. */
export function ensureHtmlExtension(path: string) {
	if (path.endsWith('.html')) return ensureLeadingSlash(path);
	path = stripLeadingAndTrailingSlashes(path);
	return path ? ensureLeadingSlash(path) + '.html' : '/index.html';
}

export function formatPath(href: string, format: 'file' | 'directory') {
	// atach base
	href = format === 'file' ? fileWithBase(href) : pathWithBase(href);
	// add html extension
	href = format === 'file' ? ensureHtmlExtension(href) : stripExtension(href);
	return href;
}
```

Then users could import it and pass the `format`:

```js
import project from 'virtual:starlight/project-context';
import { formatPath } from './path';

formatPath(path, project.build.format);
```

**Comment by kevinzunigacuellar** [comment]
I've created a new file for the format function and added `trailingSlash` as an option, given its relevance to the project configuration.

There's some ambiguity regarding the expected behavior for `trailingSlash: "ignore"`. Should we establish a convention similar to the existing one for adding trailing slashes to links? How might this convention apply when the build format is set to "file," or should we disregard it entirely?

In my view, adopting a convention would simplify the comparison of values for "isCurrent" in the sidebar. The convention I propose is as follows:

- `format: "directory", trailingSlash: "ignore"`: Retain the trailing slash (similar to the existing setup).
  - `"/reference"` becomes `"/reference/"`
  - `"/reference.html"` becomes `"/reference/"`

- `format: "file", trailingSlash: "ignore"`: Omit the trailing slash.
  - `"/reference"` becomes `"/reference.html"`
  - `"/reference/"` becomes `"/reference.html"`

Establishing this convention would enhance consistency and aid in comparing values for "isCurrent" in the sidebar, offering a clearer and more predictable structure for the project. What do you think? @delucis @HiDeoo

File: packages/starlight/utils/format-path.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+};
+
+export function formatPath(href: string, opts?: FormatPathOptions) {
+	const options = {
+		format: opts?.format ?? 'directory',
+		trailingSlash: opts?.trailingSlash ?? 'ignore',
+	};
+
+	const formatStrategy = formatStrategies[options.format];
+	const trailingSlashStrategy = trailingSlashStrategies[options.trailingSlash];
```
Comment (delucis) [inline packages/starlight/utils/format-path.ts:35]: Nit, but in general for situations like this we use defaults in the signature. My preference is mainly based on avoiding the ambiguity of having both `opts` and `options` in scope. With this way of assigning the defaults you have a single option to consume inside the function body.

```suggestion
export function formatPath(
	href: string,
	{ format = 'directory', trailingSlash = 'ignore' }: FormatPathOptions = {}
) {
	const formatStrategy = formatStrategies[format];
	const trailingSlashStrategy = trailingSlashStrategies[trailingSlash];
```

This also switches from the nullish coalescing `??` logic to logic equivalent to `||`, but I don’t think we want to support stuff like `format: ''` or `trailingSlash: 0`, so should be equivalent?

File: packages/starlight/utils/navigation.ts [hunk 10]
```diff
@@ -341,4 +355,16 @@ function applyPrevNextLinkConfig(
...
+}
+
+/** Add '.html' extension to a path. */
+export function ensureHtmlExtension(path: string) {
+	path = stripTrailingSlash(path);
+	if (path.endsWith('.html')) return ensureLeadingSlash(path);
+
+	path = stripLeadingAndTrailingSlashes(path);
+	return path ? ensureLeadingSlash(path) + '.html' : '/index.html';
+}
```
Comment (delucis) [inline packages/starlight/utils/navigation.ts:370]: What do you think of moving these to `utils/path.ts`? When they were only used by the navigation logic, keeping them in this file made sense, but now we’re sharing them also in `utils/format-path.ts`, I think that might be the more correct place for them

File: packages/starlight/utils/navigation.ts [hunk 11]
```diff
@@ -134,8 +141,15 @@ function makeLink(
 	badge?: Badge,
 	attrs?: LinkHTMLAttributes
 ): Link {
-	if (!isAbsolute(href)) href = pathWithBase(href);
-	const isCurrent = href === ensureTrailingSlash(currentPathname);
+	href = formatPath(href, {
+		format: project.build.format,
+	});
```
Comment (delucis) [inline packages/starlight/utils/navigation.ts:146]: Should we not preserve the previous behaviour and avoid formatting absolute URLs? I would guess we shouldn’t enforce our formatting on an external URL. We’re not currently testing any sidebar entries with `{ link: "https://example.com" }`, so may have missed changes in formatting here.

```suggestion
	if (!isAbsolute(href)) {
		href = formatPath(href, { format: project.build.format });
	}
```

File: packages/starlight/utils/navigation.ts [hunk 12]
```diff
@@ -134,8 +141,15 @@ function makeLink(
...
+	currentPathname =
+		project.build.format === 'file'
+			? stripTrailingSlash(currentPathname)
+			: ensureTrailingSlash(currentPathname);
+	const isCurrent = isAbsolute(href) ? false : href === currentPathname;
```
Comment (delucis) [inline packages/starlight/utils/navigation.ts:151]: Here on the other hand it should be safe to skip the `isAbsolute()` check, as absolute URLs will never match the pathname anyway:

```suggestion
	const isCurrent = href === currentPathname;
```

File: packages/starlight/utils/navigation.ts [hunk 13]
```diff
@@ -134,8 +141,15 @@ function makeLink(
...
 ): Link {
-	if (!isAbsolute(href)) href = pathWithBase(href);
-	const isCurrent = href === ensureTrailingSlash(currentPathname);
+	href = formatPath(href, {
+		format: project.build.format,
+	});
+	currentPathname =
+		project.build.format === 'file'
+			? stripTrailingSlash(currentPathname)
+			: ensureTrailingSlash(currentPathname);
```
Comment (delucis) [inline packages/starlight/utils/navigation.ts:150]: One question here: `currentPathname` is passed from `Astro.url.pathname`, does Astro provide this already formatted according to `build.format` and `trailingSlash` settings or would we need to call `formatPath(currentPathname)`?

**Review by delucis — COMMENTED** [review]
Nice refactor @kevinzunigacuellar! Really like the maps of strategies you came up with in `utils/format-path`. Few last comments and questions from me!

File: packages/starlight/utils/navigation.ts (hunk 13 above)
Comment (kevinzunigacuellar) [reply packages/starlight/utils/navigation.ts:150]: Great question! I believe the reason it doesn't autoformat is to maintain continuity.

When someone requests `/base/reference.html/`, it forwards the URL exactly as provided.

You're correct that we might need to invoke `formatPath`. This ternary operator won't function when a trailing slash is specified. I might have to adjust `formatPath` to avoid adding the base path twice. The current implementation assumes that base is not present.

**Comment by kevinzunigacuellar** [comment]
I may have pinpointed why the test "final parameter can set custom link even if no default link existed" is failing. It seems we're passing the same value to href and currentPath. 

https://github.com/withastro/starlight/blob/0dc4dc793080a96b4f21b72ace264ba6c39d6a17/packages/starlight/utils/navigation.ts#L336

I'm a bit unsure about how to address this—perhaps we could consider getting the current link as a parameter, although I'm not entirely convinced it's necessary. The scenario where this occurs is quite rare—only if someone uses the same link twice (current path name and prev/next link).

Another approach could be to make it optional or pass an empty string. 🤔 Any ideas?

**Comment by delucis** [comment]
Hey @kevinzunigacuellar! Turned out this was fiddlier than I expected, so I spent some time getting everything polished up.

Quick summary of the changes I made:

- Wire up a project’s `trailingSlash` option with the virtual module and apply it when formatting paths.
- Move around some of the logic about when slashes are added/removed to ensure that with `trailingSlash: 'ignore'` (the Astro default) we don’t modify trailing slashes on user-specified links.
- Make sure we never add a trailing slash when `build.format` is set to `file` — I’m pretty sure `/example.html` should never be `/example.html/`.
- Fixed checks for whether a sidebar link was for the current page by always formatting the comparison links with a formatter that sets `trailingSlash: 'always'` so they will be consistent.
- Made the `formatPath()` helper easier to use in Starlight code by exporting it already configured with the project settings. Formatters with different settings can be created with the `createPathFormatter()` factory method.

Would love a quick look if you have time and thanks for the patience waiting for me on this one!

**Review by kevinzunigacuellar — COMMENTED** [review]
Thanks @delucis for the review and changes. Yes, this PR started very simple but it got very complex very quickly 😅. I left a couple comments, feel free to disregard the nit comments btw.

File: packages/starlight/__tests__/basics/format-path.test.ts [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+			{ path: '/api/v1/users', expected: '/api/v1/users' },
+			// with file extension
+			{ path: '/guides/components.html', expected: '/guides/components' },
+			// with file extension and trailing slash
+			{ path: '/guides/components.html/', expected: '/guides/components' },
+		],
+	},
+])(
+	'formatPath() with { format: $options.format, trailingSlash: $options.trailingSlash }',
+	({ options, tests }) => {
```
[the remark this answers is not in the document]
Comment (delucis) [reply packages/starlight/__tests__/basics/format-path.test.ts:100]: Not sure I understand what this means. Could you explain a little more?

File: packages/starlight/utils/createPathFormatter.ts [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+	// Handle extension
+	href = formatStrategy.handleExtension(href);
+
+	// Skip trailing slash handling for `build.format: 'file'`
+	if (format === 'file') return href;
+
+	// Handle trailing slash
+	href ||= '/';
+	href = href === '/' ? href : trailingSlashStrategy(href);
```
[the remark this answers is not in the document]
Comment (delucis) [reply packages/starlight/utils/createPathFormatter.ts:51]: It’s hit in my test, although I guess the coverage may not recognise the two branches of the conditional assignment:

test coverage report

Don’t recall exactly, but I’m pretty sure I hit a case where this was needed.

File: packages/starlight/utils/createPathFormatter.ts [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+	const trailingSlashStrategy = trailingSlashStrategies[trailingSlash];
+
+	// Add base
+	href = formatStrategy.addBase(href);
+
+	// Handle extension
+	href = formatStrategy.handleExtension(href);
+
+	// Skip trailing slash handling for `build.format: 'file'`
+	if (format === 'file') return href;
```
[the remark this answers is not in the document]
Comment (delucis) [reply packages/starlight/utils/createPathFormatter.ts:47]: I don’t think so — if Astro lets you set that, then that’s on Astro 😄

File: packages/starlight/utils/createPathFormatter.ts [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+	ensureHtmlExtension,
+	ensureTrailingSlash,
+	stripHtmlExtension,
+	stripTrailingSlash,
+} from './path';
+
+interface FormatPathOptions {
+	format?: AstroConfig['build']['format'];
+	trailingSlash?: AstroConfig['trailingSlash'];
+}
```
[the remark this answers is not in the document]
Comment (delucis) [reply packages/starlight/utils/createPathFormatter.ts:13]: Ah, maybe. `base` is it a bit different because you access it with `import.meta.env.BASE_URL` instead of via our project context virtual module. I think we can leave it for now.

File: packages/starlight/utils/createPathFormatter.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (kevinzunigacuellar) [reply packages/starlight/utils/createPathFormatter.ts:51]: I see. I commented out line 50 when running unit tests and it seem pass all the test cases. Perhaps we should add a few more edge case for when it hits line 50.

File: packages/starlight/__tests__/basics/format-path.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (delucis) [reply packages/starlight/__tests__/basics/format-path.test.ts:100]: Oh right. Yeah, this was cheating a bit just to avoid rewriting the tests too much. I think it’s fine though — it’s OK to do weird stuff in tests sometimes 😅

File: packages/starlight/__tests__/basics/format-path.test.ts [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+			// index page
+			{ path: '/', expected: '/' },
+			// with trailing slash
+			{ path: '/reference/configuration/', expected: '/reference/configuration/' },
+			// without trailing slash
+			{ path: '/api/v1/users', expected: '/api/v1/users' },
+			// with file extension
+			{ path: '/guides/components.html', expected: '/guides/components' },
+			// with file extension and trailing slash
+			{ path: '/guides/components.html/', expected: '/guides/components' },
```
[the remark this answers is not in the document]
Comment (delucis) [reply packages/starlight/__tests__/basics/format-path.test.ts:95]: 🤔 I honestly don’t know… `/guides/components.html/` is already so totally badly formatted that it shouldn’t happen. I even debated if we should handle this case or only strip `.html` from paths without a trailing slash. Because on some hosts this would actually point to `/guides/components.html/index.html` 😅 

I think I’m happy with calling this one of those “bugs” we don’t have to fix until someone actually has a use case where this is an issue.

File: packages/starlight/utils/createPathFormatter.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (delucis) [reply packages/starlight/utils/createPathFormatter.ts:51]: Thanks for testing! Maybe I just added it out of fear. I’ll commit your suggestion and we can revisit it if anyone runs into issues 🙌

**Comment by kevinzunigacuellar** [comment]
Lost Glasses, Troublesome Moment! ✅
